File: py_modules/phone_import.py
"""Short-lived LAN inbox for one explicitly confirmed guide share text."""

import hmac
import json
import re
import secrets
import socket
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer


MAX_BODY_BYTES = 8 * 1024
SUCCESS = "已发送到 Deck，待确认并下载，不代表导入成功"


class _PhoneHTTPServer(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, *args):
        self._slots = threading.BoundedSemaphore(4)
        self._connections = set()
        self._connections_lock = threading.Lock()
        super().__init__(*args)

    def get_request(self):
        request, address = super().get_request()
        with self._connections_lock:
            self._connections.add(request)
        return request, address

    def shutdown_request(self, request):
        try:
            super().shutdown_request(request)
        finally:
            with self._connections_lock:
                self._connections.discard(request)

    def close_connections(self):
        with self._connections_lock:
            connections = tuple(self._connections)
        for request in connections:
            try:
                request.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass
            self.shutdown_request(request)

    def process_request(self, request, client_address):
        if not self._slots.acquire(blocking=False):
            self.shutdown_request(request)
            return
        try:
            super().process_request(request, client_address)
        except BaseException:
            self._slots.release()
            raise

    def process_request_thread(self, request, client_address):
        try:
            super().process_request_thread(request, client_address)
        finally:
            self._slots.release()

    def handle_error(self, _request, _client_address):
        # Expiring a session can race a phone disconnect; never log request details.
        pass


class PhoneImportSession:
    def __init__(
        self, *, bind_address="0.0.0.0", hostname=None, lifetime_seconds=600
    ):
        self.id = secrets.token_hex(12)
        self._token = secrets.token_urlsafe(32)
        self._hostname = hostname or socket.gethostname().removesuffix(".local") + ".local"
        if not re.fullmatch(r"[A-Za-z0-9](?:[A-Za-z0-9.-]*[A-Za-z0-9])?", self._hostname):
            raise ValueError("Steam Deck 主机名无效")
        self._bind_address = bind_address
        self._lifetime = lifetime_seconds
        self._lock = threading.Lock()
        self._server = None
        self._thread = None
        self._timer = None
        self._closed = False
        self._text = None
        self._url = None

    def start(self):
        with self._lock:
            if self._closed:
                raise RuntimeError("手机导入会话已结束，请生成新的二维码")
            if self._server is None:
                session = self

                class Handler(BaseHTTPRequestHandler):
                    def setup(self):
                        super().setup()
                        self.connection.settimeout(3)

                    def log_message(self, *_args):
                        pass

                    def reply(self, status, text, content_type="text/plain; charset=utf-8"):
                        body = text.encode("utf-8")
                        self.send_response(status)
                        self.send_header("Content-Type", content_type)
                        self.send_header("Content-Length", str(len(body)))
                        self.send_header("Cache-Control", "no-store")
                        self.send_header("Referrer-Policy", "no-referrer")
                        self.send_header("X-Content-Type-Options", "nosniff")
                        self.send_header("Connection", "close")
                        self.send_header(
                            "Content-Security-Policy",
                            "default-src 'none'; connect-src 'self'; "
                            f"script-src 'nonce-{session._nonce}'; "
                            f"style-src 'nonce-{session._nonce}'; "
                            "base-uri 'none'; form-action 'none'; frame-ancestors 'none'",
                        )
                        self.end_headers()
                        self.wfile.write(body)

                    def valid_host(self):
                        return self.headers.get_all("Host") == [session._authority]

                    def do_GET(self):
                        if not self.valid_host():
                            self.reply(403, "请求来源不正确")
                        elif self.path != "/":
                            self.reply(404, "页面不存在")
                        elif self.headers.get("Origin") not in (None, session._origin):
                            self.reply(403, "请求来源不正确")
                        else:
                            self.reply(200, session._page(), "text/html; charset=utf-8")

                    def do_POST(self):
                        if (
                            not self.valid_host()
                            or self.headers.get_all("Origin") != [session._origin]
                            or self.headers.get_all("X-GRIP-Token") is None
                            or len(self.headers.get_all("X-GRIP-Token")) != 1
                            or not hmac.compare_digest(
                                self.headers["X-GRIP-Token"].encode("utf-8"),
                                session._token.encode("ascii"),
                            )
                        ):
                            self.reply(403, "二维码无效或请求来源不正确")
                            return
                        if self.path != "/submit":
                            self.reply(404, "页面不存在")
                            return
                        if (
                            self.headers.get_all("Content-Type") != ["application/json"]
                            or self.headers.get("Transfer-Encoding") is not None
                        ):
                            self.reply(415, "仅接收分享文字")
                            return
                        lengths = self.headers.get_all("Content-Length") or []
                        if len(lengths) != 1 or not re.fullmatch(r"[0-9]{1,10}", lengths[0]):
                            self.reply(400, "请求长度无效")
                            return
                        length = int(lengths[0])
                        if length > MAX_BODY_BYTES:
                            self.reply(413, "分享文字过长")
                            return
                        try:
                            body = self.rfile.read(length)
                            if len(body) != length:
                                raise ValueError("incomplete body")
                            data = json.loads(body)
                        except (ValueError, UnicodeError, OSError):
                            self.reply(400, "分享文字格式无效")
                            return
                        if (
                            not isinstance(data, dict)
                            or set(data) != {"text"}
                            or not isinstance(data["text"], str)
                            or not data["text"].strip()
                            or any(character in data["text"] for character in "<>\x00")
                        ):
                            self.reply(400, "请发送链接或纯文本分享内容")
                            return
                        with session._lock:
                            if session._closed or time.monotonic() >= session._deadline:
                                status, response = 410, "二维码已过期，请在 Deck 上重新生成"
                            elif session._text is not None:
                                status, response = 409, "本次内容已发送，请在 Deck 上确认"
                            else:
                                session._text = data["text"].strip()
                                status, response = 200, SUCCESS
                        self.reply(status, response)

                self._server = _PhoneHTTPServer((self._bind_address, 0), Handler)
                self._authority = f"{self._hostname}:{self._server.server_port}"
                self._origin = f"http://{self._authority}"
                self._url = f"{self._origin}/#{self._token}"
                self._nonce = secrets.token_urlsafe(16)
                self._deadline = time.monotonic() + self._lifetime
                self._expires_at = int((time.time() + self._lifetime) * 1000)
                self._thread = threading.Thread(
                    target=self._server.serve_forever,
                    kwargs={"poll_interval": 0.05},
                    name="grip-phone-import",
                    daemon=True,
                )
                self._thread.start()
                self._timer = threading.Timer(self._lifetime, self.stop)
                self._timer.daemon = True
                self._timer.start()
            return {"id": self.id, "url": self._url, "expiresAt": self._expires_at}

    def snapshot(self):
        with self._lock:
            if self._closed or self._server is None or time.monotonic() >= self._deadline:
                return {"state": "expired"}
            if self._text is not None:
                return {"state": "submitted", "text": self._text}
            return {"state": "waiting"}

    def stop(self):
        with self._lock:
            self._closed = True
            server, self._server = self._server, None
            self._text = None
            if self._timer:
                self._timer.cancel()
        if server:
            server.shutdown()
            server.close_connections()
            server.server_close()
            if self._thread:
                self._thread.join(timeout=2)

    def _page(self):
        return """<!doctype html><html lang="zh-CN"><meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>发送指南到 GRIP</title><style nonce="NONCE">
body{font:17px system-ui;background:#142434;color:#fff;margin:24px;max-width:640px}
textarea,button{box-sizing:border-box;width:100%;font:inherit;border-radius:8px;padding:12px}
textarea{min-height:180px;margin:12px 0}button{background:#66c0f4;color:#101b24;border:0}
p{line-height:1.5}button:disabled{opacity:.5}
</style><h1>发送指南到 Steam Deck</h1>
<p>粘贴小黑盒分享文字。这里只发送链接，仍需在 Deck 上确认并保存。</p>
<label for="text">分享文字或链接</label><textarea id="text" maxlength="8000"></textarea>
<button id="send" type="button">发送到 Deck</button><p id="status" role="status"></p>
<script nonce="NONCE">
const token = location.hash.slice(1);
history.replaceState(null, '', location.pathname);
const button = document.getElementById('send'), status = document.getElementById('status');
if (!token) { button.disabled = true; status.textContent = '请扫描 Deck 上的完整二维码'; }
button.onclick = async () => {
  button.disabled = true;
  status.textContent = '正在发送…';
  try {
    const response = await fetch('/submit', {method:'POST',
      headers:{'Content-Type':'application/json','X-GRIP-Token':token},
      body:JSON.stringify({text:document.getElementById('text').value})});
    status.textContent = await response.text();
    button.disabled = response.ok || response.status === 409 || response.status === 410;
  } catch (_) {
    status.textContent = '连接失败，请确认手机与 Deck 在同一可信局域网，并重新扫码';
    button.disabled = false;
  }
};
</script></html>""".replace("NONCE", self._nonce)

File: backend/src/phone_import.rs
//! Explicit, short-lived LAN inbox. A submission is text, never an imported guide.

use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};
use serde_json::{Value, json};
use std::collections::HashMap;
use std::io::{self, Read, Write};
use std::net::{Shutdown, TcpListener, TcpStream};
use std::sync::{Arc, Condvar, Mutex};
use std::thread::{self, JoinHandle};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use subtle::ConstantTimeEq;

use crate::lock;

const MAX_BODY_BYTES: usize = 8 * 1024;
const MAX_HEADER_BYTES: usize = 16 * 1024;
const MAX_CONNECTIONS: usize = 4;
const IDLE_TIMEOUT: Duration = Duration::from_secs(3);
const ACCEPT_POLL: Duration = Duration::from_millis(50);
const LIFETIME: Duration = Duration::from_secs(600);
const SUCCESS: &str = "已发送到 Deck，待确认并下载，不代表导入成功";

#[derive(Default)]
struct State {
    closed: bool,
    text: Option<String>,
    connections: HashMap<u64, TcpStream>,
}

struct Shared {
    state: Mutex<State>,
    wake: Condvar,
    deadline: Instant,
    authority: String,
    origin: String,
    token: String,
    nonce: String,
    page: String,
}

impl Shared {
    fn close(&self) {
        let mut state = lock(&self.state);
        state.closed = true;
        state.text = None;
        for connection in state.connections.values() {
            let _ = connection.shutdown(Shutdown::Both);
        }
        self.wake.notify_all();
    }
}

pub struct PhoneImportSession {
    id: String,
    url: String,
    expires_at: u64,
    shared: Arc<Shared>,
    listener: Option<JoinHandle<()>>,
}

impl PhoneImportSession {
    #[cfg(test)]
    pub(crate) fn start_loopback(lifetime: Duration) -> io::Result<Self> {
        Self::start_at("127.0.0.1:0", "localhost", lifetime)
    }

    /// Nothing listens until this explicit call. The URL uses the Deck's mDNS name.
    pub fn start() -> io::Result<Self> {
        let mut hostname = [0_u8; 256];
        // SAFETY: gethostname receives a valid writable buffer and its exact length.
        if unsafe { libc::gethostname(hostname.as_mut_ptr().cast(), hostname.len()) } != 0 {
            return Err(io::Error::last_os_error());
        }
        let end = hostname
            .iter()
            .position(|byte| *byte == 0)
            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Steam Deck 主机名无效"))?;
        let hostname = std::str::from_utf8(&hostname[..end])
            .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "Steam Deck 主机名无效"))?;
        Self::start_at(
            "0.0.0.0:0",
            &format!(
                "{}.local",
                hostname.strip_suffix(".local").unwrap_or(hostname)
            ),
            LIFETIME,
        )
    }

    fn start_at(address: &str, hostname: &str, lifetime: Duration) -> io::Result<Self> {
        if hostname.is_empty()
            || !hostname.as_bytes()[0].is_ascii_alphanumeric()
            || !hostname.as_bytes()[hostname.len() - 1].is_ascii_alphanumeric()
            || !hostname
                .bytes()
                .all(|byte| byte.is_ascii_alphanumeric() || b".-".contains(&byte))
        {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "Steam Deck 主机名无效",
            ));
        }
        let id = random_bytes::<12>()?
            .iter()
            .map(|byte| format!("{byte:02x}"))
            .collect();
        let token = URL_SAFE_NO_PAD.encode(random_bytes::<32>()?);
        let nonce = URL_SAFE_NO_PAD.encode(random_bytes::<16>()?);
        let listener = TcpListener::bind(address)?;
        listener.set_nonblocking(true)?;
        let authority = format!("{hostname}:{}", listener.local_addr()?.port());
        let origin = format!("http://{authority}");
        let url = format!("{origin}/#{token}");
        let expires_at = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(io::Error::other)?
            .as_millis() as u64
            + lifetime.as_millis() as u64;
        let shared = Arc::new(Shared {
            state: Mutex::new(State::default()),
            wake: Condvar::new(),
            deadline: Instant::now() + lifetime,
            authority,
            origin,
            token,
            page: include_str!("../assets/phone-import.html").replace("NONCE", &nonce),
            nonce,
        });
        let worker_shared = shared.clone();
        let worker = thread::Builder::new()
            .name("grip-phone-import".into())
            .spawn(move || serve(listener, worker_shared))?;
        Ok(Self {
            id,
            url,
            expires_at,
            shared,
            listener: Some(worker),
        })
    }

    pub fn id(&self) -> &str {
        &self.id
    }

    pub fn info(&self) -> Value {
        json!({"id": self.id, "url": self.url, "expiresAt": self.expires_at})
    }

    pub fn snapshot(&self) -> Value {
        let state = lock(&self.shared.state);
        if state.closed || Instant::now() >= self.shared.deadline {
            json!({"state": "expired"})
        } else if let Some(text) = &state.text {
            json!({"state": "submitted", "text": text})
        } else {
            json!({"state": "waiting"})
        }
    }

    pub fn stop(&mut self) {
        self.shared.close();
        if let Some(listener) = self.listener.take() {
            let _ = listener.join();
        }
    }
}

impl Drop for PhoneImportSession {
    fn drop(&mut self) {
        self.stop();
    }
}

fn random_bytes<const N: usize>() -> io::Result<[u8; N]> {
    let mut bytes = [0; N];
    getrandom::getrandom(&mut bytes).map_err(|error| io::Error::other(error.to_string()))?;
    Ok(bytes)
}

struct Connection {
    id: u64,
    stream: TcpStream,
    shared: Arc<Shared>,
}

impl Drop for Connection {
    fn drop(&mut self) {
        let _ = self.stream.shutdown(Shutdown::Both);
        lock(&self.shared.state).connections.remove(&self.id);
    }
}

fn serve(listener: TcpListener, shared: Arc<Shared>) {
    let mut workers: Vec<JoinHandle<()>> = Vec::new();
    let mut next_id = 0;
    loop {
        let state = lock(&shared.state);
        if state.closed || Instant::now() >= shared.deadline {
            break;
        }
        drop(state);
        let mut index = 0;
        while index < workers.len() {
            if workers[index].is_finished() {
                let _ = workers.swap_remove(index).join();
            } else {
                index += 1;
            }
        }
        match listener.accept() {
            Ok((stream, _)) => {
                let Ok(tracked) = stream.try_clone() else {
                    continue;
                };
                if stream.set_nonblocking(false).is_err()
                    || stream.set_read_timeout(Some(IDLE_TIMEOUT)).is_err()
                    || stream.set_write_timeout(Some(IDLE_TIMEOUT)).is_err()
                {
                    continue;
                }
                let mut state = lock(&shared.state);
                if state.closed
                    || Instant::now() >= shared.deadline
                    || state.connections.len() >= MAX_CONNECTIONS
                {
                    let _ = stream.shutdown(Shutdown::Both);
                    continue;
                }
                next_id += 1;
                state.connections.insert(next_id, tracked);
                drop(state);
                let mut connection = Connection {
                    id: next_id,
                    stream,
                    shared: shared.clone(),
                };
                if let Ok(worker) = thread::Builder::new()
                    .name("grip-phone-request".into())
                    .spawn(move || {
                        let _ = handle(&mut connection.stream, &connection.shared);
                    })
                {
                    workers.push(worker);
                }
            }
            Err(error) if error.kind() == io::ErrorKind::WouldBlock => {
                let state = lock(&shared.state);
                if state.closed {
                    break;
                }
                let wait =
                    ACCEPT_POLL.min(shared.deadline.saturating_duration_since(Instant::now()));
                let _ = shared.wake.wait_timeout(state, wait);
            }
            Err(error) if error.kind() == io::ErrorKind::Interrupted => continue,
            Err(_) => break,
        }
    }
    shared.close();
    drop(listener);
    for worker in workers {
        let _ = worker.join();
    }
}

fn single_header<'a>(headers: &[httparse::Header<'a>], name: &str) -> Option<&'a [u8]> {
    let mut matches = headers
        .iter()
        .filter(|header| header.name.eq_ignore_ascii_case(name));
    let value = matches.next()?.value;
    matches.next().is_none().then_some(value)
}

fn handle(stream: &mut TcpStream, shared: &Shared) -> io::Result<()> {
    let mut input = Vec::new();
    let body_offset = loop {
        let mut headers = [httparse::EMPTY_HEADER; 64];
        match httparse::Request::new(&mut headers).parse(&input) {
            Ok(httparse::Status::Complete(offset)) => break offset,
            Err(_) => return reply(stream, shared, 400, "请求格式无效", false),
            Ok(httparse::Status::Partial) => {}
        }
        if input.len() >= MAX_HEADER_BYTES {
            return reply(stream, shared, 431, "请求头过长", false);
        }
        let mut chunk = [0; 1024];
        let available = chunk.len().min(MAX_HEADER_BYTES - input.len());
        let count = stream.read(&mut chunk[..available])?;
        if count == 0 {
            return Ok(());
        }
        input.extend_from_slice(&chunk[..count]);
    };
    let mut headers = [httparse::EMPTY_HEADER; 64];
    let mut request = httparse::Request::new(&mut headers);
    request
        .parse(&input)
        .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "请求格式无效"))?;
    let valid_host = single_header(request.headers, "Host") == Some(shared.authority.as_bytes());
    if request.method == Some("GET") {
        let origin = single_header(request.headers, "Origin");
        let no_origin = !request
            .headers
            .iter()
            .any(|header| header.name.eq_ignore_ascii_case("Origin"));
        return if !valid_host || (!no_origin && origin != Some(shared.origin.as_bytes())) {
            reply(stream, shared, 403, "请求来源不正确", false)
        } else if request.path != Some("/") {
            reply(stream, shared, 404, "页面不存在", false)
        } else {
            reply(stream, shared, 200, &shared.page, true)
        };
    }
    if request.method != Some("POST") {
        return reply(stream, shared, 501, "不支持的请求方法", false);
    }
    let token = single_header(request.headers, "X-GRIP-Token").unwrap_or_default();
    if !valid_host
        || single_header(request.headers, "Origin") != Some(shared.origin.as_bytes())
        || !bool::from(token.ct_eq(shared.token.as_bytes()))
    {
        return reply(stream, shared, 403, "二维码无效或请求来源不正确", false);
    }
    if request.path != Some("/submit") {
        return reply(stream, shared, 404, "页面不存在", false);
    }
    if single_header(request.headers, "Content-Type") != Some(b"application/json")
        || request
            .headers
            .iter()
            .any(|header| header.name.eq_ignore_ascii_case("Transfer-Encoding"))
    {
        return reply(stream, shared, 415, "仅接收分享文字", false);
    }
    let length = single_header(request.headers, "Content-Length")
        .filter(|value| (1..=10).contains(&value.len()) && value.iter().all(u8::is_ascii_digit))
        .and_then(|value| std::str::from_utf8(value).ok())
        .and_then(|value| value.parse::<usize>().ok());
    let Some(length) = length else {
        return reply(stream, shared, 400, "请求长度无效", false);
    };
    if length > MAX_BODY_BYTES {
        return reply(stream, shared, 413, "分享文字过长", false);
    }
    let mut body = vec![0; length];
    let buffered = length.min(input.len() - body_offset);
    body[..buffered].copy_from_slice(&input[body_offset..body_offset + buffered]);
    if stream.read_exact(&mut body[buffered..]).is_err() {
        return reply(stream, shared, 400, "分享文字格式无效", false);
    }
    let data = serde_json::from_slice::<Value>(&body).ok();
    let text = data
        .as_ref()
        .and_then(Value::as_object)
        .filter(|object| object.len() == 1)
        .and_then(|object| object.get("text"))
        .and_then(Value::as_str)
        .filter(|text| !text.trim().is_empty() && !text.contains(['<', '>', '\0']));
    let Some(text) = text else {
        return reply(stream, shared, 400, "请发送链接或纯文本分享内容", false);
    };
    let (status, message) = {
        let mut state = lock(&shared.state);
        if state.closed || Instant::now() >= shared.deadline {
            (410, "二维码已过期，请在 Deck 上重新生成")
        } else if state.text.is_some() {
            (409, "本次内容已发送，请在 Deck 上确认")
        } else {
            state.text = Some(text.trim().to_owned());
            (200, SUCCESS)
        }
    };
    reply(stream, shared, status, message, false)
}

fn reply(
    stream: &mut TcpStream,
    shared: &Shared,
    status: u16,
    body: &str,
    html: bool,
) -> io::Result<()> {
    let content_type = if html {
        "text/html; charset=utf-8"
    } else {
        "text/plain; charset=utf-8"
    };
    let response = format!(
        "HTTP/1.1 {status} Response\r\nContent-Type: {content_type}\r\nContent-Length: {}\r\nCache-Control: no-store\r\nReferrer-Policy: no-referrer\r\nX-Content-Type-Options: nosniff\r\nConnection: close\r\nContent-Security-Policy: default-src 'none'; connect-src 'self'; script-src 'nonce-{}'; style-src 'nonce-{}'; base-uri 'none'; form-action 'none'; frame-ancestors 'none'\r\n\r\n{body}",
        body.len(),
        shared.nonce,
        shared.nonce
    );
    stream.write_all(response.as_bytes())
}

#[cfg(test)]
#[path = "phone_import_tests.rs"]
mod tests;
